Make Table.join use ordered joins on OrderedRVIterator

patrick-schultz opened the pull request:
This PR is to get Table.join to stop using Spark's join using Annotations, instead using ordered joins directly with region values.

Putting the new join into the IR will be a future PR.

I've given significant thought to how to make joins work making zero assumptions about partition keys. I'm convinced this is the right thing to do, and that with the right foundations it would just work without much extra code. I started to build those foundations in this PR, but some more work is needed to remove all partition key preconditions.

patrick-schultz commented:
cc @cseed (say that three times fast!)

tpoterba commented on src/main/scala/is/hail/annotations/OrderedRVIterator.scala:
```diff
@@ -5,6 +5,15 @@ import is.hail.utils._
+  def restrictToPKInterval(interval: Interval): Iterator[RegionValue] = {
+    val ur = new UnsafeRow(t.rowType, null, 0)
+    val pk = new KeyedRow(ur, t.kRowFieldIdx)
+    iterator.filter( rv => {
```
prefer curly braces around the entire thing instead of parens and curly braces together

tpoterba commented on src/main/scala/is/hail/annotations/UnsafeRow.scala:
```diff
@@ -301,3 +301,14 @@ class UnsafeRow(var t: TBaseStruct,
 }
+
+class KeyedRow(var row: Row, keyFields: Array[Int]) extends Row {
+  def this(row: Row) = this(row, Array.range(0, row.size))
```
this constructor doesn't seem used, can we delete it?

patrick-schultz replied:
Oh, yes. That was from when I didn't have `KeyedRow` extending `Row`.

tpoterba commented on src/main/scala/is/hail/expr/types/TStruct.scala:
```diff
@@ -90,6 +90,13 @@ final case class TStruct(fields: IndexedSeq[Field], override val required: Boole
   }
 
+  def isIsomorphicTo(other: TStruct): Boolean =
+    size == other.size && isPrefixOf(other)
```
isn't this exactly `isOfType(other)`

patrick-schultz replied:
`isOfType` compares names also. There are places, such as testing for joinability, where you want to ignore the top-level names of the struct.

Since the lower-level names still matter, I'm not really happy with the name `isIsomorphicTo`. Another option would be to put a field on `TStruct` which gives the corresponding `Tuple`, and use `isOfType` on the `Tuple`s.

tpoterba replied:
ahhh, I see. This is fine, then.

tpoterba commented on src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:
```diff
@@ -0,0 +1,69 @@
+  * needed. No assumption should need to be made about partition keys, but currently
+  * assumes old partition key type is a prefix of the new partition key type.
+  */
+class RepartitionedOrderedRDD2(
```
RepartitionedOrderedRVD?

patrick-schultz replied:
This is an `RDD[RegionValue]`, not an `RVD`.

tpoterba commented on src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:
```diff
@@ -0,0 +1,69 @@
+
+case class RepartitionedOrderedRDD2Partition(
+    index: Int,
+    parents: Seq[Partition],
```
can this be an array? It's probably a List as implemented here.

patrick-schultz replied:
I was copying from older code, and wasn't sure why some places used `Array` and some `Seq`. `Array` should be fine.

tpoterba replied:
probably because Spark uses Seq everywhere :)

tpoterba commented on src/main/scala/is/hail/table/Table.scala:
```diff
@@ -626,70 +626,104 @@ class Table(val hc: HailContext, val ir: TableIR) {
-      case "outer" => rddLeft.fullOuterJoin(rddRight).map { case (k, (l, r)) => merger(k, l.orNull, r.orNull) }
-      case _ => fatal("Invalid join type specified. Choose one of `left', `right', `inner', `outer'")
+    val left = this.rvd match {
+      case ordered: OrderedRVD => ordered
```
what about the case that it's ordered but keyed by the wrong thing? I'm having a bit of trouble reasoning about that case right now. Could it happen?

tpoterba replied:
ah, keyBy does a horrible thing right now. so maybe it can't?

tpoterba replied:
let's add an assertion, at least.

patrick-schultz replied:
I think we need to decide what keys mean in different places. One thing I think would be reasonable is
* Keys on `Table`, both in Python and in Scala, are purely metadata used to determine how to join.
* Keys on `OrderedRVD` are a way of recording the ordering invariant satisfied by the underlying `RDD`.

If this is the case, then I need to rewrite this. When I wrote it, I think I was assuming the `Table` keys and the `OrderedRVD` keys had to be the same. Now I understand `Table`s are never ordered currently. We should probably make a reasonable choice of what keys mean and document it.

tpoterba commented on src/main/scala/is/hail/table/Table.scala:
```diff
@@ -626,70 +626,104 @@ class Table(val hc: HailContext, val ir: TableIR) {
+      joinType,
+      rvMerger,
+      new OrderedRVDType(left.typ.partitionKey, left.typ.key, newSignature))
+    copy2(rvd = joinedRVD, signature = newSignature, key = key)
```
don't need the key=key I think

tpoterba commented on src/test/scala/is/hail/rvd/OrderedRVDPartitionerSuite.scala:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  // @Test def testGetPartitionPKWithSmallerKeys() {
+  //   assert(partitioner.getPartitionPK(Row(2)) == 0)
```
meant to be uncommented?

patrick-schultz replied:
No, this is a test that I think should be made to pass, but currently doesn't. This is where I need the general comparison of arbitrary length tuples. I could remove this, or add a comment saying to enable the test when we can.

patrick-schultz commented:
We should discuss the struct ordering in person. I think there are orderings that can be defined on the space of all tuples (since the names don't matter) of arbitrary lengths, which are very helpful in working with changing keys and partition keys. In principle, it should be easy to repartition an OrderedRVD with a longer partition key to a partitioner with a shorter partition key, but currently that doesn't look simple to do. I tried to lay the groundwork here to make that trivial.

patrick-schultz commented:
Speaking of not understanding what keys mean, I found what looks to me like a bug, but I'm not sure. `OrderedRVD.downcastToPK` creates an `OrderedRVD` for which `typ.kType` is different from `partitioner.kType`. It's triggering the assert I made in `RepartitionedOrderedRDD2` that says the new key must be a prefix of the old, to ensure that no sorting needs to be done.

I want to make join keys parameters of `OrderedRVD.join`, allowing them to be different from the partitioner keys. I was putting that off for a later PR, but now I think I might need to do that to fix this.

patrick-schultz commented:
I addressed most of your comments. I also fixed the `downcastToPK` problem by getting rid of it, instead adding a `KeyedOrderedRVD` which has a join key in addition to an ordering key.

patrick-schultz commented:
I moved table join to the IR, and it was such a tiny change I just added it to this PR. I can separate it back out if you'd prefer.

tpoterba commented on src/main/scala/is/hail/rvd/UnpartitionedRVD.scala:
```diff
@@ -49,4 +49,26 @@ class UnpartitionedRVD(val rowType: TStruct, val rdd: RDD[RegionValue]) extends
+      }
+    }
+
+    OrderedRVD.shuffle(ordType, newPartitioner, filtered)
```
we might want to consider using the coerce strategy with `newPartitioner` as a hint partitioner. Maybe not, though.

patrick-schultz replied:
`coerce` felt too high level for this, but I'm open to discussion. This method says "give me back an `OrderedRVD` with exactly this partitioner, dropping any data that fall outside the given partitions." So the partitioner argument is more than a hint.

The other factor is that `coerce` does a pass over the data collecting statistics, and detecting if it was already ordered. My feeling was that this method should be an explicit "just do a shuffle". Probably collecting statistics and making choices based on the results should be explicit in the IR, not built into the methods the IR calls. Maybe @cseed has thoughts?

patrick-schultz commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -95,42 +80,57 @@ class OrderedRVD private(
+
+    require(ordType.rowType == typ.rowType)
+    require(ordType.kType isPrefixOf typ.kType)
+//    require(newPartitioner.kType isIsomorphicTo ordType.kType)
```
It's frustrating because this should be an invariant satisfied by all `OrderedRVD`s. But I didn't see any simple way to make it pass, and I'm planning on a refactoring that removes `kType` from `OrderedRVDPartitioner`, getting rid of the redundancy. I'll just delete this.
